### PR TITLE
[CSM] Add camera API

### PR DESCRIPTION
**Pull request by bigfoot547**
This PR adds an api for the camera class.
TODO:
 * [x] Populate with functions
 * [x] Add documentation

*Yay! I learned to use git!*

### CODE REVIEW

File: src/camera.cpp [hunk 1]
```diff
@@ -126,6 +127,10 @@ bool Camera::successfullyCreated(std::string &error_message)
 	} else {
 		error_message.clear();
 	}
+	if (g_settings->getBool("enable_client_modding")) {
```
Comment (nerzhul) [inline src/camera.cpp]: can you add an empty line upper ?

File: src/camera.cpp [hunk 2]
```diff
@@ -126,6 +127,10 @@ bool Camera::successfullyCreated(std::string &error_message)
...
+		ClientScripting *script = m_client->getScript();
```
Comment (nerzhul) [inline src/camera.cpp]: no need for a temp variable

File: src/camera.h [hunk 3]
```diff
@@ -160,7 +160,12 @@ class Camera
 		else
 			m_camera_mode = CAMERA_MODE_FIRST;
 	}
-
+	
+	// Set the current camera mode
+	void setCameraMode(CameraMode mode) {
```
Comment (nerzhul) [inline src/camera.h]: brace on wrong line

File: src/script/lua_api/l_camera.cpp [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+	Camera *camera = getobject(L, 1);
+	if (!camera)
+		return 0;
+	
+	lua_pushnumber(L, (int)camera->getCameraMode());
+	
+	return 1;
+}
+
+int LuaCamera::l_get_fov_x(lua_State *L)
```
Comment (nerzhul) [inline src/script/lua_api/l_camera.cpp]: why not have a single call get_fov which returns {x = , y = } ?

Comment (bigfoot547) [inline src/script/lua_api/l_camera.cpp]: I will return a table of `{x = get_fov_x, y = get_fov_y, actual = get_fov}` .

File: src/camera.h [hunk 5]
```diff
@@ -160,7 +160,13 @@ class Camera
 		else
 			m_camera_mode = CAMERA_MODE_FIRST;
 	}
-
+	
+	// Set the current camera mode
+	void setCameraMode(CameraMode mode)
```
Comment (SmallJoker) [inline src/camera.h]: `inline`

File: doc/client_lua_api.md [hunk 6]
```diff
@@ -782,6 +782,41 @@ An interface to manipulate minimap on client UI
...
+    * Returns with same syntax as above
+* `get_fov()`
+    * Returns:
+    
+```lua
+    {
+        x = number,
+	y = number,
+	actual = number,
+	max = number
```
Comment (SmallJoker) [inline doc/client_lua_api.md]: Formatting.

File: doc/client_lua_api.md [hunk 7]
```diff
@@ -782,6 +782,41 @@ An interface to manipulate minimap on client UI
 * `set_shape(shape)`: Sets the minimap shape. (0 = square, 1 = round)
 * `get_shape()`: Gets the minimap shape. (0 = square, 1 = round)
 
+### Camera
+An interface to get or set information about the camera and cameranode.
+Please do not try to access the reference until a second or so after the client
```
Comment (SmallJoker) [inline doc/client_lua_api.md]: `or so`. Accuracy could be better here.
Just add a notice that the reference will be nil until the camera is initialized.

File: src/script/clientscripting.cpp [hunk 8]
```diff
@@ -71,10 +72,17 @@ void ClientScripting::InitializeModApi(lua_State *L, int top)
...
 void ClientScripting::on_client_ready(LocalPlayer *localplayer)
 {
 	lua_State *L = getStack();
 	LuaLocalPlayer::create(L, localplayer);
 }
+
+void ClientScripting::on_camera_ready(Camera *camera)
+{
+	lua_State *L = getStack();
+	LuaCamera::create(L, camera);
```
Comment (SmallJoker) [inline src/script/clientscripting.cpp]: Superflous temporary variable.
`LuaCamera::create(getStack(), camera);`

File: doc/client_lua_api.md (hunk 6 above)
Comment (bigfoot547) [inline doc/client_lua_api.md]: What do you mean?

File: doc/client_lua_api.md [hunk 9]
```diff
@@ -782,6 +782,40 @@ An interface to manipulate minimap on client UI
 * `set_shape(shape)`: Sets the minimap shape. (0 = square, 1 = round)
 * `get_shape()`: Gets the minimap shape. (0 = square, 1 = round)
 
+### Camera
+An interface to get or set information about the camera and cameranode.
+Please do not try to access the reference until the camera is initialized, otherwise the reference will be nil.
```
Comment (nerzhul) [inline doc/client_lua_api.md]: You should precise minetest.camera else we don't know what you are talking about

**Comment by bigfoot547** [comment]
Squashed! Merge ready!

File: src/script/lua_api/l_camera.h [hunk 10]
```diff
@@ -0,0 +1,47 @@
+#ifndef L_CAMERA_H
+#define L_CAMERA_H
+
+#include "l_base.h"
+#include "camera.h"
```
Comment (nerzhul) [inline src/script/lua_api/l_camera.h]: why include camera.h you already declared class Camera later ?

Comment (bigfoot547) [inline src/script/lua_api/l_camera.h]: It wouldn't compile without it.

File: doc/client_lua_api.md (hunk 6 above)
Comment (SmallJoker) [inline doc/client_lua_api.md]: Indents should be like here: https://github.com/minetest/minetest/pull/5609/files#diff-927bb329aa13096bcf9123248a422bc2R864

File: src/script/lua_api/CMakeLists.txt [hunk 11]
```diff
@@ -29,4 +29,5 @@ set(client_SCRIPT_LUA_API_SRCS
 	${CMAKE_CURRENT_SOURCE_DIR}/l_storage.cpp
 	${CMAKE_CURRENT_SOURCE_DIR}/l_sound.cpp
 	${CMAKE_CURRENT_SOURCE_DIR}/l_localplayer.cpp
+	${CMAKE_CURRENT_SOURCE_DIR}/l_camera.cpp
```
Comment (SmallJoker) [inline src/script/lua_api/CMakeLists.txt]: Update Android's list aswell: `build/android/jni/Android.mk`

Comment (bigfoot547) [inline src/script/lua_api/CMakeLists.txt]: Done!

File: doc/client_lua_api.md (hunk 6 above)
Comment (bigfoot547) [inline doc/client_lua_api.md]: Done!

**Comment by bigfoot547** [comment]
@nerzhul @SmallJoker Please review

File: src/script/lua_api/l_camera.cpp [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+	int coretable = lua_gettop(L);
+	
+	lua_pushvalue(L, camera_object);
+	lua_setfield(L, coretable, "camera");
+}
+
+int LuaCamera::l_set_camera_mode(lua_State *L)
+{
+	Camera *camera = getobject(L, 1);
+	GenericCAO *playercao = getClient(L)->getEnv().getLocalPlayer()->getCAO();
```
Comment (nerzhul) [inline src/script/lua_api/l_camera.cpp]: you don't test this pointer (sanity_check)

File: src/script/lua_api/l_camera.cpp [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+	if (!camera)
+		return 0;
+	
+	push_v3s16(L, camera->getOffset());
+	return 1;
+}
+
+int LuaCamera::l_get_look_dir(lua_State *L)
+{
+	LocalPlayer *player = getClient(L)->getEnv().getLocalPlayer();
```
Comment (nerzhul) [inline src/script/lua_api/l_camera.cpp]: you don't test this pointer (sanity_check)

File: src/script/lua_api/l_camera.cpp (hunk 12 above)
Comment (bigfoot547) [inline src/script/lua_api/l_camera.cpp]: Done!

File: src/script/lua_api/l_camera.cpp (hunk 13 above)
Comment (bigfoot547) [inline src/script/lua_api/l_camera.cpp]: Done!

**Comment by bigfoot547** [comment]
Rebased!

**Comment by SmallJoker** [comment]
Please don't use the rebase feature from GitHub, it messes up the whole commit history. Instead, fetch minetest/minetest for updates and use the rebase feature in your Git client.

**Comment by bigfoot547** [comment]
@nerzhul @SmallJoker  Please review

File: src/script/lua_api/l_camera.h (hunk 10 above)
Comment (nerzhul) [inline src/script/lua_api/l_camera.h]: please remove include , class has been declared, and include camera.h in cpp file

**Comment by bigfoot547** [comment]
Ok @nerzhul can you please review one more time?
Then we can finally merge this!

**Comment by bigfoot547** [comment]
Yay! Thanks! :smiley:

**Pull request merged**